src/test/java/rx/ObservableTests.java

```diff
@@ -1174,4 +1175,304 @@ public Object call(OnSubscribe<Object> onSubscribe) {
             }});
     }
+    
+    @Test
+    public void nullOnSubscribe() {
+        Observable<Integer> source = Observable.create((OnSubscribe<Integer>)null);
+        
+        try {
+            source.subscribe();
+            fail("Should have thrown IllegalStateException");
+        } catch (IllegalStateException ex) {
+            // expected
+        }
+    }
+    
+    @Test
+    public void nullObserver() {
+        Observable<Integer> source = Observable.just(1);
+        
+        try {
+            source.subscribe((Observer<Integer>)null);
+            fail("Should have thrown IllegalStateException");
+        } catch (NullPointerException ex) {
+            // expected
+        }
+    }
+
+    @Test
+    public void nullSubscriber() {
+        Observable<Integer> source = Observable.just(1);
+        
+        try {
+            source.subscribe((Subscriber<Integer>)null);
+            fail("Should have thrown IllegalStateException");
+        } catch (IllegalArgumentException ex) {
+            // expected
+        }
+    }
+
+    @SuppressWarnings("deprecation")
+    @Test
+    public void testCacheHint() throws InterruptedException {
+        final AtomicInteger counter = new AtomicInteger();
+        Observable<String> o = Observable.create(new OnSubscribe<String>() {
+
+            @Override
+            public void call(final Subscriber<? super String> observer) {
+                new Thread(new Runnable() {
+
+                    @Override
+                    public void run() {
+                        counter.incrementAndGet();
+                        observer.onNext("one");
+                        observer.onCompleted();
+                    }
+                }).start();
+            }
+        }).cache(1);
+
+        // we then expect the following 2 subscriptions to get that same value
+        final CountDownLatch latch = new CountDownLatch(2);
+
+        // subscribe once
+        o.subscribe(new Action1<String>() {
+
+            @Override
+            public void call(String v) {
+                assertEquals("one", v);
+                latch.countDown();
+            }
+        });
+
+        // subscribe again
+        o.subscribe(new Action1<String>() {
+
+            @Override
+            public void call(String v) {
+                assertEquals("one", v);
+                latch.countDown();
+            }
+        });
+
+        if (!latch.await(1000, TimeUnit.MILLISECONDS)) {
+            fail("subscriptions did not receive values");
+        }
+        assertEquals(1, counter.get());
+    }
+    
+    @Test
+    public void subscribeWithNull() {
+        Action1<Integer> onNext = Actions.empty();
+        Action1<Throwable> onError = Actions.empty();
+        Action0 onCompleted = Actions.empty();
+        try {
+            Observable.just(1).subscribe((Action1<Integer>)null);
+            fail("Should have thrown IllegalArgumentException");
```

Comment:
👍 for `fail()` without a message + `expected` for exception name.

---

>optimizing for the failure case seems like the wrong thing to do

Can't completely agree with that, 2 reasons why I (and probably others) love AssertJ and similar libraries instead of JUnit's assertions:

1. Fluent API with specialization, makes assertions human-readable
2. Really good fail messages with pointing to what was expected and what has actually happened in a very readable manner